### PR TITLE
Add service check for libvirtd

### PR DESCRIPTION
**Pull request by coolgw**
Add service check for libvirtd

- Related ticket: https://progress.opensuse.org/issues/94285
- Needles:na
- Verification run: 
  opensuse: https://progress.opensuse.org/issues/94285

### CODE REVIEW

File: lib/services/libvirtd.pm [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+use base 'opensusebasetest';
+use testapi;
+use utils;
+use strict;
+use warnings;
+
+#default guest vm params
+our %guest_params = (
+    'name'        => 'nested_vm_test',
+    'ram' => '',
```
Comment (rfan1) [inline lib/services/libvirtd.pm:24]: ram != memory? right?

File: lib/services/libvirtd.pm [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    zypper_call("lr -u");
+    for (my $i = 1; $i < $repo_max_num + 1; $i = $i +1) {
+        zypper_call("rr 1");
+    }
+    script_run("zypper lr -u");
+}
+sub install_service {
+    if (get_var("DISTRI") eq "opensuse") {
+        remove_repo();
+        my $version = script_output("grep VERSION= /etc/os-release | cut -d'=' -f2 | cut -d' ' -f1 | sed 's/\"//g'");
```
Comment (rfan1) [inline lib/services/libvirtd.pm:45]: Wei, not a big deal, but you can take a look at "lsb_release -a" cmd
suse@localhost:~> lsb_release -a
LSB Version:	n/a
Distributor ID:	openSUSE
Description:	openSUSE Leap 15.2
Release:	15.2
Codename:	n/a

File: lib/services/libvirtd.pm [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        check_guest_status(%guest_params , 'status'=>'running');
+	shutdown_guest(%guest_params) if get_var("NESTED_VM_DOWN");
+    }
+
+    common_service_action($pkg, $type, 'is-enabled');
+    common_service_action($pkg, $type, 'is-active');
+
+    if ($stage eq 'after') {
+        script_run("virsh list --all");
+        check_guest_status(%guest_params , 'status'=>'shut');
```
Comment (rfan1) [inline lib/services/libvirtd.pm:128]: Can you please add some comments why we check status=shut based on our offline discussion?

File: tests/console/nested_vm.pm [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# SUSE's openQA tests
+#
+# Copyright © 2019 SUSE LLC
```
Comment (rfan1) [inline tests/console/nested_vm.pm:3]: s/2019/2021/g

File: tests/console/nested_vm.pm [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        my %hash = (stage => 'before', service_type => 'Systemd', srv_pkg_name => 'libvirtd');
+        services::libvirtd::full_libvirtd_check(%hash);
+        set_var('NESTED_VM_CHECK_BEFORE_MIGRATION', 1);
+        enter_cmd "reboot";
+    }
+}
+
+sub post_run_hook {
+    # Do nothing
+}
```
Comment (rfan1) [inline tests/console/nested_vm.pm:37]: Remove these lines?

Comment (coolgw) [reply tests/console/nested_vm.pm:37]: This is used to override default post_run_hook:)

File: lib/services/libvirtd.pm [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    'boot'  => 'cdrom',
+);
+
+sub remove_repo {
+    my $repo_max_num = script_output("zypper lr | tail -1 | cut -d' ' -f1");
+    zypper_call("lr -u");
+    for (my $i = 1; $i < $repo_max_num + 1; $i = $i +1) {
+        zypper_call("rr 1");
+    }
+    script_run("zypper lr -u");
```
Comment (foursixnine) [inline lib/services/libvirtd.pm:40]: why script run here, instead of zypper_call?

File: lib/services/libvirtd.pm [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+}
+sub install_service {
+    if (get_var("DISTRI") eq "opensuse") {
+        remove_repo();
+        my $version = script_output("grep VERSION= /etc/os-release | cut -d'=' -f2 | cut -d' ' -f1 | sed 's/\"//g'");
+        zypper_call("ar http://download.opensuse.org/distribution/leap/$version/repo/oss/ main");
+        zypper_call("ar http://download.opensuse.org/update/leap/$version/oss/ mainupdate");
```
Comment (foursixnine) [inline lib/services/libvirtd.pm:47]: I guess this is meant to be a hardcoded test for leap right?, if so would you mind adding this for TW?

File: lib/services/libvirtd.pm [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        zypper_call("ref");
+        zypper_call("lr -u");
+    }
+    zypper_call('in libvirt-daemon virt-install');
+    zypper_call('in -t pattern kvm_server kvm_tools');
```
Comment (foursixnine) [inline lib/services/libvirtd.pm:52]: Shouldn't `kvm_server` and `kvm_tools` be enough here to get libvirt-daemon, virt-install and friends installed?

File: tests/console/nested_vm.pm [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Package: nest_vm
+# Summary: Basics libvirtd test
```
Comment (foursixnine) [inline tests/console/nested_vm.pm:11]: Would you adding a bit more of the intention of the test:

e.g: ` Test libvirt pre and post migration, with a nested VM running or shutdown, performance of the nested vm is irrelevant as long as the service is able to complete the upgrade` or similar to what the schedule says :)

**Review by foursixnine — COMMENTED** [review]
@coolgw Awesome! Thanks a lot for taking my suggestion into account!

File: lib/services/libvirtd.pm (hunk 1 above)
Comment (coolgw) [reply lib/services/libvirtd.pm:24]: yes, ram is for ram parameter of virsh command

File: lib/services/libvirtd.pm (hunk 6 above)
Comment (coolgw) [reply lib/services/libvirtd.pm:40]: zypper_call will failed if can not find any repo, so have to use script_run

File: lib/services/libvirtd.pm (hunk 7 above)
Comment (coolgw) [reply lib/services/libvirtd.pm:47]: I only find upgrade case from leap->TW in openqa.opensuse.org
I can not find any case for TW to TW, do we have this kind of check?

Comment (coolgw) [reply lib/services/libvirtd.pm:47]: @foursixnine ?

Comment (foursixnine) [reply lib/services/libvirtd.pm:47]: Was just a question, I don't think we have TW to TW for now we need to introduce within qe-core [poo#16618](https://progress.opensuse.org/issues/16618) for that.

**Review by lemon-suse — APPROVED** [review]

File: lib/services/libvirtd.pm (hunk 7 above)
Comment (coolgw) [reply lib/services/libvirtd.pm:47]: Notify me or create ticket for me once  TW to TW is ready, then we can integrate it.

**Review by foursixnine — APPROVED** [review]

**Pull request merged**